### PR TITLE
[NFC] Refactor name parsing into `parseDeclNameRef()`

### PR DESCRIPTION
**Pull request by beccadax**
The parser currently has two DeclNameRef-parsing functions, `parseUnqualifiedDeclName()` and `parseUnqualifiedDeclBaseName()`, which take four undifferentiated `bool`s to control their behavior. I will soon need yet another flag, so this is getting a bit out of hand.

This PR replaces them with a `parseDeclNameRef()` function and a flag enum to control these behaviors. This should improve the readability of existing code and help accommodate future flag additions.

### CODE REVIEW

**Comment by brentdax** [comment]
@swift-ci please test

File: lib/Parse/ParseExpr.cpp [hunk 1]
```diff
@@ -2115,95 +2197,26 @@ DeclNameRef Parser::parseUnqualifiedDeclBaseName(
...
-    SourceLoc argLoc;
-    parseOptionalArgumentLabel(argName, argLoc);
-    if (argLoc.isValid()) {
-      argumentLabels.push_back(argName);
-      argumentLabelLocs.push_back(argLoc);
-      continue;
-    }
+  bool hadArgList = tryParseArgLabelList(*this, flags, lparenLoc,
+                                         argumentLabels, argumentLabelLocs,
+                                         rparenLoc);
```
Comment (rintaro) [inline lib/Parse/ParseExpr.cpp]: This unconditionally tries to parse argument labels part. Don't we need to differentiate former `parseUnqualifiedDeclBaseName()` call?
For example, for `Foo(foo:bar:)` at type position, we should not parse it as valid `IdentTypeRepr`, but should stop parsing before the argument part (and diagnose it).

Comment (brentdax) [reply lib/Parse/ParseExpr.cpp]: The first thing `tryParseArgLabelList` does is check if the `AllowCompoundNames` flag is set and bail out if it isn’t. I could hoist that check into this method, though.

Comment (rintaro) [reply lib/Parse/ParseExpr.cpp]: Ah I see. Thanks for explanation!

**Review by rintaro — APPROVED** [review]
Nice cleanup. Thanks!

**Comment by brentdax** [comment]
@swift-ci please test linux platform

**Comment by brentdax** [comment]
Possibly a flaky test.

@swift-ci please test linux platform

**Pull request merged**